### PR TITLE
make assertion in backchaining more explicitly checked

### PR DESCRIPTION
**Pull request by ronuchit**

### CODE REVIEW

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py [hunk 1]
```diff
@@ -390,3 +378,19 @@ def _finalize_pnad_side_predicates(pnad: PartialNSRTAndDatastore) -> None:
             for atom in next_atoms - segment.final_atoms:
                 side_predicates.add(atom.predicate)
         pnad.op = pnad.op.copy_with(side_predicates=side_predicates)
+
+    def _assert_all_data_in_exactly_one_datastore(
+            self, pnads: List[PartialNSRTAndDatastore]) -> None:
+        """Assert that every demo datapoint appears in exactly one datastore
```
Comment (tomsilver) [inline src/nsrt_learning/strips_learning/gen_to_spec_learner.py:384]: maybe i'm being dum but i don't see how this checks for EXACTLY one?

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py [hunk 2]
```diff
@@ -390,3 +378,19 @@ def _finalize_pnad_side_predicates(pnad: PartialNSRTAndDatastore) -> None:
...
+        among the given PNADs' datastores."""
+        all_segs_in_data_lst = [
+            seg for pnad in pnads for seg, _ in pnad.datastore
+        ]
+        all_segs_in_data = set(all_segs_in_data_lst)
```
[the remark this answers is not in the document]
Comment (ronuchit) [reply src/nsrt_learning/strips_learning/gen_to_spec_learner.py:389]: i got the idea for this after coming upon a thing called `self._segment_to_grounding`. is that okay? i bet it's just hashing based on the object itself which is what we want

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py (hunk 1 above)
Comment (ronuchit) [reply src/nsrt_learning/strips_learning/gen_to_spec_learner.py:384]: the `len(x) == len(set(x))` check

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomsilver) [reply src/nsrt_learning/strips_learning/gen_to_spec_learner.py:389]: yeah cool

File: src/nsrt_learning/strips_learning/gen_to_spec_learner.py (hunk 1 above)
Comment (tomsilver) [reply src/nsrt_learning/strips_learning/gen_to_spec_learner.py:384]: yup was being dumb cool

**Review by tomsilver — APPROVED** [review]

**Review by NishanthJKumar — APPROVED** [review]
LGTM! Thanks for doing this; it's much cleaner and obvious what we're checking now!

**Pull request merged**